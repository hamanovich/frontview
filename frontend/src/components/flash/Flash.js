import React from 'react';
import { func, shape, string } from 'prop-types';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';
import styled from 'styled-components';
import classNames from 'classnames';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';

const enhance = compose(
  lifecycle({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

const FlashItem = styled(Alert)`
  display: flex;
  justify-content: space-between;
  position: fixed;
  z-index: 10;
  left: 50%;
  transform: translateX(-50%);
  top: 52px;
  align-items: center;
  min-width: 290px;
  max-width: 90%;

  button {
    margin-left: 1.5rem;
  }

  + .alert {
    margin-top: 4rem;

    + .alert {
      margin-top: 12rem;
    }
  }
`;

export const Flash = ({ close, message }) => (
  <FlashItem
    variant={classNames({
      danger: message.type === 'error',
      success: message.type === 'success',
      warning: message.type === 'warn',
    })}>
    <MarkdownRenderer markdown={message.text} />
    <Button onClick={close} className="close">
      <FontAwesome name="times" />
    </Button>
  </FlashItem>
);

Flash.propTypes = {
  message: shape({
    type: string,
    text: string,
  }).isRequired,
  close: func.isRequired,
};

export default enhance(Flash);
