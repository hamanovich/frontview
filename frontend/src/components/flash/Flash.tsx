import React, { FC, ReactNode } from 'react';
import FontAwesome from 'react-fontawesome';
import MarkdownRenderer from 'react-markdown-renderer';

import compose from 'recompose/compose';
import lifecycle from 'recompose/lifecycle';

import Button from 'react-bootstrap/Button';

import { FlashProps } from './models';

import { FlashItem } from './style';

const enhance = compose<FlashProps, FlashProps>(
  lifecycle<FlashProps, {}>({
    componentDidMount() {
      setTimeout(this.props.close, 4000);
    },
  }),
);

export const Flash: FC<FlashProps> = ({ close, message }) => {
  const variant =
    message.type === 'error'
      ? 'danger'
      : message.type === 'success'
      ? 'success'
      : 'warning';
  return (
    <FlashItem variant={variant}>
      <MarkdownRenderer markdown={message.text} />
      <Button onClick={close} className="close">
        <FontAwesome name="times" />
      </Button>
    </FlashItem>
  );
};

export default enhance(Flash);
