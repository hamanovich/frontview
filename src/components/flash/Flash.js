import React from 'react';
import PropTypes from 'prop-types';
import FontAwesome from 'react-fontawesome';

import Alert from 'react-bootstrap/lib/Alert';
import Button from 'react-bootstrap/lib/Button';

const Flash = ({ close, message }) => (
  <Alert bsStyle={message.type === 'error' ? 'danger' : 'success'}>
    {message.text}
    <Button onClick={close} className="close">
      <FontAwesome name="times" />
    </Button>
  </Alert>
);

Flash.propTypes = {
  message: PropTypes.shape({
    type: PropTypes.string,
    text: PropTypes.string
  }).isRequired,
  close: PropTypes.func.isRequired
};

export default Flash;
