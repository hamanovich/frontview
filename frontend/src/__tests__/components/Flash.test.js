import React from 'react';
import { shallow } from 'enzyme';

import Flash from '../../components/flash/Flash';

describe('<Flash/>', () => {
  const props = {
    message: {
      type: 'error',
      text: 'Error',
    },
    close: jest.fn(),
  };

  const component = shallow(<Flash {...props} />);

  it('renders <Flash /> component', () => {
    expect(component).toMatchSnapshot();
  });

  it('updates via props', () => {
    component.setProps({
      message: { type: 'success', text: 'Success' },
    });

    expect(component.find('span').text()).toEqual('Success');
    expect(component).toMatchSnapshot();
  });

  describe('when clicking close button', () => {
    beforeEach(() => component.find('Button').simulate('click'));

    it('invokes the close callback', () => {
      expect(props.close).toHaveBeenCalled();
    });
  });
});
