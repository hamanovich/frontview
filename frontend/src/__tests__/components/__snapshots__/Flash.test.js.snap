// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flash/> renders <Flash /> component 1`] = `
<Styled(Alert)
  variant="danger"
>
  <MarkdownRenderer
    markdown="Error"
  />
  <Button
    active={false}
    className="close"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
    variant="primary"
  >
    <FontAwesome
      name="times"
    />
  </Button>
</Styled(Alert)>
`;

exports[`<Flash/> updates via props 1`] = `
<Styled(Alert)
  variant="success"
>
  <MarkdownRenderer
    markdown="Success"
  />
  <Button
    active={false}
    className="close"
    disabled={false}
    onClick={[MockFunction]}
    type="button"
    variant="primary"
  >
    <FontAwesome
      name="times"
    />
  </Button>
</Styled(Alert)>
`;
